File: book1_companion/psych_stats_with_python_companion_v0_2/scripts/r/ch07_one_way_anova.R
#!/usr/bin/env Rscript
args <- commandArgs(trailingOnly = FALSE)
script_arg <- args[grep('^--file=', args)][1]
root <- normalizePath(file.path(dirname(sub('^--file=', '', script_arg)), '..', '..'), mustWork = TRUE)
source(file.path(root, "scripts", "r", "common.R"))

d <- read.csv(file.path(root, "data", "ch07_one_way_anova.csv"))
fit <- aov(test_score ~ study_condition, data = d)
tab <- summary(fit)[[1]]
ss_between <- tab["study_condition", "Sum Sq"]; ss_within <- tab["Residuals", "Sum Sq"]
means <- tapply(d$test_score, d$study_condition, mean)
write_result(root, "ch07", "One-way ANOVA", "data/ch07_one_way_anova.csv", list(
  practice_quiz_mean = means[["practice_quiz"]], retrieval_practice_mean = means[["retrieval_practice"]],
  spaced_review_mean = means[["spaced_review"]], f_statistic = tab["study_condition", "F value"],
  df_between = tab["study_condition", "Df"], df_within = tab["Residuals", "Df"],
  p_value = tab["study_condition", "Pr(>F)"], eta_squared = ss_between / (ss_between + ss_within)
))

File: book1_companion/psych_stats_with_python_companion_v0_2/scripts/python/ch07_one_way_anova.py
#!/usr/bin/env python3
from __future__ import annotations

import pandas as pd
from scipy import stats

from common import ROOT, rounded, write_result


def main() -> None:
    data_file = "data/ch07_one_way_anova.csv"
    data = pd.read_csv(ROOT / data_file)
    groups = {name: frame["test_score"] for name, frame in data.groupby("study_condition", sort=True)}
    test = stats.f_oneway(*groups.values())
    grand_mean = data["test_score"].mean()
    ss_between = sum(len(values) * (values.mean() - grand_mean) ** 2 for values in groups.values())
    ss_within = sum(((values - values.mean()) ** 2).sum() for values in groups.values())
    fields = {
        "practice_quiz_mean": rounded(groups["practice_quiz"].mean()),
        "retrieval_practice_mean": rounded(groups["retrieval_practice"].mean()),
        "spaced_review_mean": rounded(groups["spaced_review"].mean()),
        "f_statistic": rounded(test.statistic),
        "df_between": 2,
        "df_within": int(len(data) - 3),
        "p_value": rounded(test.pvalue),
        "eta_squared": rounded(ss_between / (ss_between + ss_within)),
    }
    out = write_result("ch07", "One-way ANOVA", data_file, fields)
    print(f"wrote {out}")


if __name__ == "__main__":
    main()
